elrond-wasm/src/types/interaction/contract_call.rs

```diff
@@ -1,84 +1,115 @@
-		self.hex_data.push_argument_bytes(bytes);
+		self.arg_buffer.push_argument_bytes(bytes);
 	}
 
-	pub fn push_callback_argument_raw_bytes(&mut self, bytes: &[u8]) {
-		self.hex_data.push_argument_bytes(bytes);
-	}
+	/// If this is an ESDT call, it converts it to a regular call to ESDTTransfer.
+	/// Async calls require this step, but not `transfer_esdt_execute`.
+	fn convert_to_esdt_transfer_call(self) -> Self {
+		if !self.token.is_egld() {
+			let mut new_arg_buffer = ArgBuffer::new();
+			new_arg_buffer.push_argument_bytes(self.token.as_slice());
+			new_arg_buffer.push_argument_bytes(self.payment.to_bytes_be().as_slice());
+			new_arg_buffer.push_argument_bytes(self.endpoint_name.as_slice());
 
-	pub fn push_argument_or_exit<A: AsyncCallArg, ExitCtx: Clone>(
-		&mut self,
-		arg: A,
-		c: ExitCtx,
-		exit: fn(ExitCtx, SCError) -> !,
-	) {
-		// TODO: propagate fast exit down the serializer
-		// TODO: also expose an EncodingError instead of the SCError
-		if let Err(serialization_err) = arg.push_async_arg(&mut self.hex_data) {
-			exit(c, serialization_err);
+			ContractCall {
+				to: self.to,
+				token: TokenIdentifier::egld(),
+				payment: BigUint::zero(),
+				endpoint_name: BoxedBytes::from(ESDT_TRANSFER_STRING),
+				arg_buffer: new_arg_buffer.concat(self.arg_buffer),
+			}
+		} else {
+			self
 		}
 	}
 
-	pub fn async_call(self) -> AsyncCall<BigUint> {
+	pub fn async_call(mut self) -> AsyncCall<BigUint> {
+		self = self.convert_to_esdt_transfer_call();
 		AsyncCall {
 			to: self.to,
-			egld_payment: self.egld_payment,
-			hex_data: self.hex_data,
+			egld_payment: self.payment,
+			hex_data: HexCallDataSerializer::from_arg_buffer(
+				self.endpoint_name.as_slice(),
+				&self.arg_buffer,
+			),
 			callback_data: HexCallDataSerializer::new(&[]),
 		}
 	}
+
+	/// Produces an EGLD (or no value) transfer-execute call, no callback.
+	/// Will always result in a `transferValueExecute` call.
+	pub fn transfer_egld_execute(self) -> TransferEgldExecute<BigUint> {
+		TransferEgldExecute {
+			to: self.to,
+			egld_payment: self.payment,
+			endpoint_name: self.endpoint_name,
+			arg_buffer: self.arg_buffer,
+			gas_limit: 0,
+		}
+	}
+
+	/// Produces an ESDT transfer-execute call, no callback.
+	/// Will always result in a `transferESDTExecute` call.
+	pub fn transfer_esdt_execute(self) -> TransferEsdtExecute<BigUint> {
+		TransferEsdtExecute {
+			to: self.to,
+			token_name: self.token.into_boxed_bytes(),
+			amount: self.payment,
+			endpoint_name: self.endpoint_name,
+			arg_buffer: self.arg_buffer,
+			gas_limit: 0,
+		}
+	}
+
+	/// Produces a transfer-execute call, no callback.
+	/// Will result in either a `transferValueExecute` or a `transferESDTExecute` call, depending on input.
+	pub fn transfer_execute(self) -> TransferExecute<BigUint> {
+		TransferExecute {
+			to: self.to,
+			token: self.token,
+			amount: self.payment,
+			endpoint_name: self.endpoint_name,
+			arg_buffer: self.arg_buffer,
+			gas_limit: 0,
+		}
+	}
```

Comment:
Wouldn't it be a bit better to have a single TransferExecute struct? Having TransferEgldExecute and TransferEsdtExecute seems a bit overkill. This is entirely optional and you probably have a good reason for splitting them like this, but give it a thought, at least.

Reply:
If the choice between the hooks is done at runtime, the compiler cannot know which one will be called and both end up in the binary. The only way to separate them at compile time is to have separate structs (or another level of generics, but that's just overkill, I feel). I tried to keep the structs absolutely minimal. The relevant logic is in the conversion methods in ContractCall.